File: python/src/multifrost/transport.py
"""
FILE: python/src/multifrost/transport.py
PURPOSE: Own the live WebSocket transport used by both caller and service peers.
OWNS: WebSocket connection lifecycle, register ack handling, pending request tracking, and inbound frame routing.
EXPORTS: TransportHandlers, WebSocketTransport.
DOCS: docs/spec.md; agent_chat/python_v5_api_surface_2026-03-25.md
"""

from __future__ import annotations

import asyncio
import contextlib
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any

from websockets.asyncio.client import connect as ws_connect
from websockets.exceptions import ConnectionClosed, WebSocketException

from .errors import ErrorOrigin, RegistrationError, TransportError
from .frame import FrameParts, decode_body, decode_frame, encode_body, encode_frame
from .protocol import (
    Envelope,
    ErrorBody,
    PeerClass,
    RegisterAckBody,
    RegisterBody,
    build_disconnect_envelope,
    build_register_envelope,
    call,
    disconnect,
    heartbeat,
    query,
    response,
)
from .protocol import (
    error as kind_error,
)

FrameHandler = Callable[[FrameParts], Awaitable[None] | None]


@dataclass(slots=True)
class TransportHandlers:
    on_response: FrameHandler | None = None
    on_error: FrameHandler | None = None
    on_call: FrameHandler | None = None
    on_query: FrameHandler | None = None
    on_heartbeat: FrameHandler | None = None
    on_disconnect: FrameHandler | None = None


@dataclass(slots=True)
class _PendingRequest:
    future: asyncio.Future[FrameParts]


class WebSocketTransport:
    def __init__(
        self,
        *,
        peer_id: str,
        peer_class: PeerClass,
        router_endpoint: str,
        handlers: TransportHandlers | None = None,
        register_timeout: float = 10.0,
        request_timeout: float | None = 30.0,
    ) -> None:
        self.peer_id = peer_id
        self.peer_class = peer_class
        self.router_endpoint = router_endpoint
        self.handlers = handlers or TransportHandlers()
        self.register_timeout = register_timeout
        self.request_timeout = request_timeout
        self._websocket: Any = None
        self._receiver_task: asyncio.Task[None] | None = None
        self._pending: dict[str, _PendingRequest] = {}
        self._send_lock = asyncio.Lock()
        self._closed_event = asyncio.Event()
        self._active = False
        self._closed = False
        self._close_reason: str | None = None

    @property
    def active(self) -> bool:
        return self._active and not self._closed

    async def start(self) -> RegisterAckBody:
        if self._websocket is not None:
            raise TransportError("transport already started")

        try:
            self._websocket = await ws_connect(
                self.router_endpoint,
                open_timeout=5.0,
                close_timeout=1.0,
                ping_interval=None,
                max_size=None,
            )
        except (OSError, TimeoutError, WebSocketException) as err:
            raise TransportError(
                f"failed to connect websocket to {self.router_endpoint}",
                cause=err,
            ) from err

        try:
            ack = await self._send_register_and_wait_ack()
        except Exception:
            await self.close(graceful=False)
            raise

        self._closed_event.clear()
        self._active = True
        self._receiver_task = asyncio.create_task(self._receive_loop())
        return ack

    async def request(
        self,
        envelope: Envelope,
        body_bytes: bytes,
        *,
        timeout: float | None = None,
    ) -> FrameParts:
        if not self.active:
            raise TransportError("transport is not active")

        loop = asyncio.get_running_loop()
        future: asyncio.Future[FrameParts] = loop.create_future()
        self._pending[envelope.msg_id] = _PendingRequest(future=future)
        frame_bytes = encode_frame(envelope, body_bytes)
        try:
            async with self._send_lock:
                await self._websocket.send(frame_bytes)
        except Exception as err:
            self._pending.pop(envelope.msg_id, None)
            raise TransportError("failed to send websocket frame", cause=err) from err

        effective_timeout = self.request_timeout if timeout is None else timeout
        try:
            if effective_timeout is None:
                return await future
            return await asyncio.wait_for(future, timeout=effective_timeout)
        except asyncio.TimeoutError as err:
            self._pending.pop(envelope.msg_id, None)
            raise TransportError(
                f"timed out waiting for response to msg_id={envelope.msg_id}",
                cause=err,
            ) from err

    async def send(self, envelope: Envelope, body_bytes: bytes) -> None:
        if not self.active:
            raise TransportError("transport is not active")
        frame_bytes = encode_frame(envelope, body_bytes)
        try:
            async with self._send_lock:
                await self._websocket.send(frame_bytes)
        except Exception as err:
            raise TransportError("failed to send websocket frame", cause=err) from err

    async def close(self, *, graceful: bool = True) -> None:
        if self._closed:
            return
        self._closed = True
        self._active = False

        if graceful and self._websocket is not None:
            with contextlib.suppress(Exception):
                disconnect_envelope = build_disconnect_envelope(self.peer_id)
                await self._websocket.send(encode_frame(disconnect_envelope, b""))

        if self._websocket is not None:
            with contextlib.suppress(Exception):
                await self._websocket.close()

        if self._receiver_task is not None:
            self._receiver_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._receiver_task

        self._fail_pending("transport closed")
        self._websocket = None
        self._closed_event.set()

    async def wait_closed(self) -> None:
        await self._closed_event.wait()

    async def _send_register_and_wait_ack(self) -> RegisterAckBody:
        register_body = RegisterBody(peer_id=self.peer_id, peer_class=self.peer_class)
        envelope = build_register_envelope(self.peer_id)
        await self._websocket.send(encode_frame(envelope, encode_body(register_body)))

        try:
            raw_message = await asyncio.wait_for(
                self._websocket.recv(), timeout=self.register_timeout
            )
        except asyncio.TimeoutError as err:
            raise RegistrationError(
                f"timed out waiting for register ack at {self.router_endpoint}",
                reason="timeout",
            ) from err
        except ConnectionClosed as err:
            raise RegistrationError(
                "router closed the websocket during registration",
                reason="router closed",
            ) from err
        except Exception as err:
            raise RegistrationError(
                f"failed while waiting for register ack: {err}",
                reason=str(err),
            ) from err

        if not isinstance(raw_message, (bytes, bytearray, memoryview)):
            raise RegistrationError(
                f"expected binary register ack, got {type(raw_message)!r}",
                reason="non-binary register reply",
            )

        frame = decode_frame(bytes(raw_message))
        if frame.envelope.kind == response:
            return self._decode_register_ack(frame)
        if frame.envelope.kind == kind_error:
            body = self._decode_error_body(frame)
            raise RegistrationError(
                body.message,
                reason=body.message,
                origin=ErrorOrigin.ROUTER,
            )
        raise RegistrationError(
            f"unexpected register reply kind: {frame.envelope.kind}",
            reason=f"unexpected kind: {frame.envelope.kind}",
        )

    def _decode_register_ack(self, frame: FrameParts) -> RegisterAckBody:
        body_data = decode_body(frame.body_bytes)
        if not isinstance(body_data, dict):
            raise RegistrationError(
                f"invalid register ack body type: {type(body_data)!r}",
                reason="invalid register ack body",
            )
        ack = RegisterAckBody.from_mapping(body_data)
        if not ack.accepted:
            raise RegistrationError(
                ack.reason or "router rejected registration",
                reason=ack.reason,
                origin=ErrorOrigin.ROUTER,
            )
        return ack

    def _decode_error_body(self, frame: FrameParts) -> ErrorBody:
        body_data = decode_body(frame.body_bytes)
        if not isinstance(body_data, dict):
            raise TransportError(
                f"invalid error body type: {type(body_data)!r}"
            )
        return ErrorBody.from_mapping(body_data)

    async def _receive_loop(self) -> None:
        try:
            while True:
                raw_message = await self._websocket.recv()
                if not isinstance(raw_message, (bytes, bytearray, memoryview)):
                    raise TransportError(
                        f"expected binary websocket frame, got {type(raw_message)!r}"
                    )

                frame = decode_frame(bytes(raw_message))
                kind = frame.envelope.kind

                if kind == response:
                    if await self._resolve_pending(frame):
                        continue
                    await self._dispatch(self.handlers.on_response, frame)
                    continue

                if kind == kind_error:
                    if await self._resolve_pending(frame):
                        continue
                    await self._dispatch(self.handlers.on_error, frame)
                    continue

                if kind == call:
                    await self._dispatch(self.handlers.on_call, frame)
                    continue

                if kind == query:
                    await self._dispatch(self.handlers.on_query, frame)
                    continue

                if kind == heartbeat:
                    await self._dispatch(self.handlers.on_heartbeat, frame)
                    continue

                if kind == disconnect:
                    await self._dispatch(self.handlers.on_disconnect, frame)
                    break

                raise TransportError(f"unexpected message kind: {kind}")
        except ConnectionClosed:
            pass
        except Exception as err:
            self._close_reason = str(err)
        finally:
            self._active = False
            self._closed = True
            self._fail_pending(self._close_reason or "transport closed")
            self._closed_event.set()

    async def _resolve_pending(self, frame: FrameParts) -> bool:
        pending = self._pending.pop(frame.envelope.msg_id, None)
        if pending is None:
            return False
        if not pending.future.done():
            pending.future.set_result(frame)
        return True

    async def _dispatch(self, handler: FrameHandler | None, frame: FrameParts) -> None:
        if handler is None:
            return
        result = handler(frame)
        if asyncio.iscoroutine(result):
            await result

    def _fail_pending(self, reason: str) -> None:
        error = TransportError(reason)
        for pending in self._pending.values():
            if not pending.future.done():
                pending.future.set_exception(error)
        self._pending.clear()


__all__ = ["TransportHandlers", "WebSocketTransport"]

File: javascript/src/transport.ts
/**
 * FILE: javascript/src/transport.ts
 * PURPOSE: Own the live WebSocket peer transport and route binary frames by v5 message kind.
 * OWNS: Socket lifecycle, binary frame send/receive, register handshake, and kind-based dispatch.
 * EXPORTS: WebSocketTransport, TransportCallbacks, RoutedFrame, TransportConnectOptions
 * DOCS: docs/spec.md
 */

import { randomUUID } from "crypto";
import WebSocket, { RawData } from "ws";
import {
    decodeBody,
    decodeEnvelope,
    decodeFrame,
    encodeBody,
    encodeFrame,
} from "./frame.js";
import {
    caller,
    Envelope,
    EnvelopeKind,
    KIND_CALL,
    KIND_DISCONNECT,
    KIND_ERROR,
    KIND_HEARTBEAT,
    KIND_QUERY,
    KIND_REGISTER,
    KIND_RESPONSE,
    PeerClass,
    RegisterAckBody,
    RegisterBody,
    service,
    validateEnvelopeFields,
} from "./protocol.js";
import {
    BootstrapFailureError,
    ProtocolError,
    RegistrationFailureError,
    TransportFailureError,
} from "./errors.js";

export interface RoutedFrame {
    envelope: Envelope;
    bodyBytes: Uint8Array;
}

export interface TransportCallbacks {
    response?: (frame: RoutedFrame) => void | Promise<void>;
    error?: (frame: RoutedFrame) => void | Promise<void>;
    query?: (frame: RoutedFrame) => void | Promise<void>;
    call?: (frame: RoutedFrame) => void | Promise<void>;
    heartbeat?: (frame: RoutedFrame) => void | Promise<void>;
    disconnect?: (frame: RoutedFrame) => void | Promise<void>;
    close?: (error?: Error) => void | Promise<void>;
}

export interface TransportConnectOptions {
    endpoint: string;
    peerId: string;
    peerClass: PeerClass;
    callbacks?: TransportCallbacks;
    registerTimeoutMs?: number;
}

function isBinaryLike(data: RawData): boolean {
    return typeof data !== "string";
}

function rawDataToBytes(data: RawData): Uint8Array {
    if (Buffer.isBuffer(data)) {
        return data;
    }

    if (Array.isArray(data)) {
        return Buffer.concat(data.map(part => (Buffer.isBuffer(part) ? part : Buffer.from(part))));
    }

    return Buffer.from(data);
}

function toFrame(raw: RawData): RoutedFrame {
    const bytes = rawDataToBytes(raw);
    const parts = decodeFrame(bytes);
    return {
        envelope: decodeEnvelope(parts.envelopeBytes),
        bodyBytes: parts.bodyBytes,
    };
}

function nowSeconds(): number {
    return Date.now() / 1000;
}

export class WebSocketTransport {
    private readonly socket: WebSocket;
    private readonly endpoint: string;
    private readonly peerId: string;
    private readonly peerClass: PeerClass;
    private readonly callbacks: TransportCallbacks;
    private readonly registerTimeoutMs: number;
    private closed = false;
    private active = false;
    private readonly closedPromise: Promise<void>;
    private readonly closedResolve: () => void;

    private constructor(
        socket: WebSocket,
        options: TransportConnectOptions,
        closedPromise: Promise<void>,
        closedResolve: () => void
    ) {
        this.socket = socket;
        this.endpoint = options.endpoint;
        this.peerId = options.peerId;
        this.peerClass = options.peerClass;
        this.callbacks = options.callbacks ?? {};
        this.registerTimeoutMs = options.registerTimeoutMs ?? 5_000;
        this.closedPromise = closedPromise;
        this.closedResolve = closedResolve;
    }

    static async connect(options: TransportConnectOptions): Promise<WebSocketTransport> {
        const socket = new WebSocket(options.endpoint);
        let closedResolve!: () => void;
        const closedPromise = new Promise<void>(resolve => {
            closedResolve = resolve;
        });

        const transport = new WebSocketTransport(
            socket,
            options,
            closedPromise,
            closedResolve
        );

        await transport.waitForRegister();
        transport.active = true;
        return transport;
    }

    get isActive(): boolean {
        return this.active && !this.closed;
    }

    get isClosed(): boolean {
        return this.closed;
    }

    get endpointUrl(): string {
        return this.endpoint;
    }

    get peer(): { peerId: string; peerClass: PeerClass } {
        return {
            peerId: this.peerId,
            peerClass: this.peerClass,
        };
    }

    async send(envelope: Envelope, bodyBytes: Uint8Array): Promise<void> {
        if (!this.isActive) {
            throw new TransportFailureError("transport is not active");
        }

        const frame = encodeFrame(validateEnvelopeFields(envelope), bodyBytes);
        await this.sendBytes(frame);
    }

    async sendBody(envelope: Envelope, body: unknown): Promise<void> {
        await this.send(envelope, encodeBody(body));
    }

    async close(): Promise<void> {
        if (this.closed) {
            return;
        }

        this.closed = true;
        try {
            if (this.socket.readyState === WebSocket.OPEN || this.socket.readyState === WebSocket.CONNECTING) {
                this.socket.close();
            }
        } finally {
            this.closedResolve();
        }
    }

    waitForClose(): Promise<void> {
        return this.closedPromise;
    }

    private async waitForRegister(): Promise<void> {
        const registerMsgId = randomUUID();
        const registerEnvelope: Envelope = {
            v: 5,
            kind: KIND_REGISTER,
            msg_id: registerMsgId,
            from: this.peerId,
            to: "router",
            ts: nowSeconds(),
        };
        const registerBody: RegisterBody = {
            peer_id: this.peerId,
            class: this.peerClass,
        };
        const registerAck = new Promise<void>((resolve, reject) => {
            const timeout = setTimeout(() => {
                cleanup();
                reject(
                    new RegistrationFailureError(
                        `registration timed out after ${this.registerTimeoutMs}ms`
                    )
                );
            }, this.registerTimeoutMs);

            const cleanup = () => clearTimeout(timeout);

            const onError = (error: Error) => {
                cleanup();
                reject(new TransportFailureError(error.message, { cause: error }));
            };

            const onClose = () => {
                cleanup();
                if (!this.active) {
                    reject(new TransportFailureError("socket closed before registration completed"));
                } else {
                    resolve();
                }
            };

            this.socket.once("error", onError);
            this.socket.once("close", onClose);

            const onMessage = (data: RawData, isBinary: boolean) => {
                if (!isBinary || !isBinaryLike(data)) {
                    cleanup();
                    reject(new ProtocolError("router sent a non-binary register reply"));
                    return;
                }

                try {
                    const frame = toFrame(data);
                    if (frame.envelope.msg_id !== registerMsgId) {
                        cleanup();
                        reject(
                            new RegistrationFailureError(
                                `unexpected message during registration: ${frame.envelope.kind}`
                            )
                        );
                        return;
                    }

                    if (frame.envelope.kind !== KIND_RESPONSE) {
                        cleanup();
                        reject(
                            new RegistrationFailureError(
                                `unexpected register reply kind: ${frame.envelope.kind}`
                            )
                        );
                        return;
                    }

                    const ack = decodeBody<RegisterAckBody>(frame.bodyBytes);
                    if (!ack.accepted) {
                        cleanup();
                        reject(
                            new RegistrationFailureError(ack.reason ?? "router rejected registration")
                        );
                        return;
                    }

                    cleanup();
                    this.socket.off("error", onError);
                    this.socket.off("close", onClose);
                    this.socket.off("message", onMessage);
                    resolve();
                } catch (error) {
                    cleanup();
                    reject(
                        error instanceof Error
                            ? error
                            : new RegistrationFailureError(String(error))
                    );
                }
            };

            this.socket.on("message", onMessage);
        });

        const open = new Promise<void>((resolve, reject) => {
            this.socket.once("open", () => {
                void this.sendBytes(encodeFrame(registerEnvelope, encodeBody(registerBody)))
                    .then(() => resolve())
                    .catch(reject);
            });
            this.socket.once("error", (error: Error) => {
                reject(new TransportFailureError(error.message, { cause: error }));
            });
        });

        await open;
        await registerAck;
        this.attachRuntimeListeners();
    }

    private attachRuntimeListeners(): void {
        this.socket.on("message", (data: RawData, isBinary: boolean) => {
            if (!isBinary || !isBinaryLike(data)) {
                this.failTransport(new ProtocolError("received non-binary websocket message"));
                return;
            }

            let frame: RoutedFrame;
            try {
                frame = toFrame(data);
            } catch (error) {
                this.failTransport(error instanceof Error ? error : new ProtocolError(String(error)));
                return;
            }

            this.dispatchFrame(frame);
        });

        this.socket.on("close", () => {
            this.closed = true;
            this.active = false;
            this.closedResolve();
            void Promise.resolve(this.callbacks.close?.()).catch(() => undefined);
        });

        this.socket.on("error", (error: Error) => {
            this.failTransport(new TransportFailureError(error.message, { cause: error }));
        });
    }

    private dispatchFrame(frame: RoutedFrame): void {
        const handler = (() => {
            switch (frame.envelope.kind) {
                case KIND_RESPONSE:
                    return this.callbacks.response;
                case KIND_ERROR:
                    return this.callbacks.error;
                case KIND_QUERY:
                    return this.callbacks.query;
                case KIND_CALL:
                    return this.callbacks.call;
                case KIND_HEARTBEAT:
                    return this.callbacks.heartbeat;
                case KIND_DISCONNECT:
                    return this.callbacks.disconnect;
                default:
                    return undefined;
            }
        })();
        if (!handler || typeof handler !== "function") {
            return;
        }

        void Promise.resolve(handler(frame)).catch(() => undefined);
    }

    private failTransport(error: Error): void {
        if (this.closed) {
            return;
        }

        this.closed = true;
        this.active = false;
        try {
            this.socket.terminate();
        } catch {
            // Ignore close failures.
        }
        this.closedResolve();
        void Promise.resolve(this.callbacks.close?.(error)).catch(() => undefined);
    }

    private async sendBytes(bytes: Uint8Array): Promise<void> {
        if (this.closed) {
            throw new TransportFailureError("transport is closed");
        }

        await new Promise<void>((resolve, reject) => {
            this.socket.send(bytes, { binary: true }, (error: Error | undefined) => {
                if (error) {
                    reject(new TransportFailureError(error.message, { cause: error }));
                } else {
                    resolve();
                }
            });
        });
    }
}
